Gems/Atom/Tools/AtomToolsFramework/Code/Source/AssetSelection/AssetSelectionComboBox.cpp

```diff
@@ -78,34 +74,108 @@ namespace AtomToolsFramework
+            }
+        }
+    }
+
+    void AssetSelectionComboBox::SetThumbnailDelayMs(AZ::u32 delay)
+    {
+        m_thumbnailDelayMs = delay;
+    }
+
     void AssetSelectionComboBox::OnCatalogAssetAdded(const AZ::Data::AssetId& assetId)
     {
         AZ::Data::AssetCatalogRequestBus::Broadcast(
             [this, assetId](AZ::Data::AssetCatalogRequests* assetCatalogRequests)
             {
-                AZ::Data::AssetInfo assetInfo = assetCatalogRequests->GetAssetInfoById(assetId);
-                if (m_filterCallback && m_filterCallback(assetInfo))
-                {
-                    addItem(
-                        GetDisplayNameFromPath(AZ::RPI::AssetUtils::GetSourcePathByAssetId(assetInfo.m_assetId).c_str()),
-                        QVariant(QString(assetInfo.m_assetId.ToString<AZStd::string>().c_str())));
-                }
+                AddAsset(assetCatalogRequests->GetAssetInfoById(assetId));
                 model()->sort(0, Qt::AscendingOrder);
             });
     }
 
-    void AssetSelectionComboBox::OnCatalogAssetRemoved(
-        [[maybe_unused]] const AZ::Data::AssetId& assetId, const AZ::Data::AssetInfo& assetInfo)
+    void AssetSelectionComboBox::OnCatalogAssetRemoved(const AZ::Data::AssetId& assetId, const AZ::Data::AssetInfo& assetInfo)
     {
         if (m_filterCallback && m_filterCallback(assetInfo))
         {
-            int index = findData(QVariant(QString(assetInfo.m_assetId.ToString<AZStd::string>().c_str())));
-            if (index >= 0 && index < count())
+            int index = findData(QVariant(QString(assetId.ToString<AZStd::string>().c_str())));
+            if (index >= 0)
             {
                 removeItem(index);
             }
         }
     }
+
+    void AssetSelectionComboBox::AddAsset(const AZ::Data::AssetInfo& assetInfo)
+    {
+        if (m_filterCallback && m_filterCallback(assetInfo))
+        {
+            addItem(
+                GetDisplayNameFromPath(AZ::RPI::AssetUtils::GetSourcePathByAssetId(assetInfo.m_assetId).c_str()),
+                QVariant(QString(assetInfo.m_assetId.ToString<AZStd::string>().c_str())));
+
+            RegisterThumbnail(assetInfo.m_assetId);
+        }
+    }
+
+    void AssetSelectionComboBox::RegisterThumbnail(const AZ::Data::AssetId& assetId)
+    {
+        if (m_thumbnailsEnabled)
+        {
+            AzToolsFramework::Thumbnailer::SharedThumbnailKey thumbnailKey =
+                MAKE_TKEY(AzToolsFramework::AssetBrowser::ProductThumbnailKey, assetId);
+            m_thumbnailKeys[assetId] = thumbnailKey;
+
+            connect(
+                thumbnailKey.data(), &AzToolsFramework::Thumbnailer::ThumbnailKey::ThumbnailUpdatedSignal, this,
+                [this, assetId]() { QueueUpdateThumbnail(assetId); });
+
+            QueueUpdateThumbnail(assetId);
+        }
+    }
+
+    void AssetSelectionComboBox::UpdateThumbnail(const AZ::Data::AssetId& assetId)
+    {
+        if (m_thumbnailsEnabled)
```

Comment:
uber-nit: early out instead of nesting?

Reply:
I could go either way. Because we require braces for even single line early returns, I think the one level of nesting looks cleaner. If it was multiple levels then I would switch it.